xdsl/transforms/lower_csl_stencil.py

```diff
@@ -0,0 +1,219 @@
+    assert (offset[0] == 0) ^ (
+        offset[1] == 0
+    ), "Expecting neighbour access in a star-shape pattern"
+    if offset[0] < 0:
+        d = csl.Direction.EAST
+    elif offset[0] > 0:
+        d = csl.Direction.WEST
+    elif offset[1] < 0:
+        d = csl.Direction.NORTH
+    elif offset[1] > 0:
+        d = csl.Direction.SOUTH
+    else:
+        raise ValueError(
+            "Invalid offset, expecting 2-dimensional star-shape neighbor access"
+        )
+    max_distance = abs(max(offset, key=abs))
+    return csl.DirectionOp(d), arith.Constant(IntegerAttr(max_distance, 16))
+
+
+def _get_module_wrapper(op: Operation) -> csl_wrapper.ModuleOp | None:
+    """
+    Return the enclosing csl_wrapper.module
+    """
+    parent_op = op.parent_op()
+    while parent_op:
+        if isinstance(parent_op, csl_wrapper.ModuleOp):
+            return parent_op
+        parent_op = parent_op.parent_op()
+    return None
+
+
+@dataclass(frozen=True)
+class LowerAccessOp(RewritePattern):
+    """
+    Replaces `csl_stencil.access` with API calls.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.AccessOp, rewriter: PatternRewriter, /):
+        if not (module_wrapper_op := _get_module_wrapper(op)):
+            return
+
+        dir_op, neighbor_op = get_dir_and_distance_ops(op)
+        rewriter.replace_matched_op(
+            [
+                neighbor_op,
+                dir_op,
+                m_call := csl.MemberCallOp(
+                    "getRecvBufDsdByNeighbor",
+                    csl.DsdType(csl.DsdKind.mem1d_dsd),
+                    module_wrapper_op.get_program_import("stencil_comms.csl"),
+                    [
+                        dir_op,
+                        neighbor_op,
+                    ],
+                ),
+                UnrealizedConversionCastOp.get([m_call], op.result_types),
+            ]
+        )
+
+
+@dataclass(frozen=True)
+class LowerApplyOp(RewritePattern):
+    """
+    Lowers csl_stencil.apply to an API call. Places the two regions in csl.funcs and
+    passes them as callbacks.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.ApplyOp, rewriter: PatternRewriter, /):
+        if not (module_wrapper_op := _get_module_wrapper(op)):
+            return
+
+        parent_func = op.parent_op()
+        while parent_func:
+            if isinstance(parent_func, func.FuncOp) or isinstance(
+                parent_func, csl.FuncOp
+            ):
+                break
+            parent_func = op.parent_op()
+        assert (
+            parent_func
+        ), "Expected csl_stencil.apply to be inside a func.func or csl.func"
+
+        # set up csl funcs
+        reduce_fn = csl.FuncOp("chunk_reduce_cb", FunctionType.from_lists([i16], []))
+        post_fn = csl.FuncOp(
+            "post_process_cb", FunctionType.from_lists([], []), Region(Block())
+        )
```

Comment:
This will cause issues if we ever have multiple `apply`s, but I don't think we support that anyway :shrug:

Reply:
fixed